libbpf-tools/tcpretrans.c

```diff
@@ -0,0 +1,359 @@
+static struct env {
+	bool verbose;
+	bool count;
+	bool lossprobe;
+	bool kprobe;
+} env = {};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'c':
+		env.count = true;
+		break;
+	case 'l':
+		env.lossprobe = true;
+		break;
+	case 'k':
+		env.kprobe = true;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void print_count_ipv4(int map_fd)
+{
+	static struct ipv4_flow_key keys[MAX_ENTRIES];
+	__u32 value_size = sizeof(__u64);
+	__u32 key_size = sizeof(keys[0]);
+	static struct ipv4_flow_key zero;
+	static __u64 counts[MAX_ENTRIES];
+	char s[INET_ADDRSTRLEN];
+	char d[INET_ADDRSTRLEN];
+	__u32 i, n = MAX_ENTRIES;
+	struct in_addr src;
+	struct in_addr dst;
+	char *ep_fmt = "[%s]#%d";
+	char remote[INET_ADDRSTRLEN + 8];
+	char local[INET_ADDRSTRLEN + 8];
+
+	if (dump_hash(map_fd, keys, key_size, counts, value_size, &n, &zero)) {
+		warn("dump_hash: %s", strerror(errno));
+		return;
+	}
+
+	for (i = 0; i < n; i++) {
+		src.s_addr = keys[i].saddr;
+		dst.s_addr = keys[i].daddr;
+		sprintf(local, ep_fmt, inet_ntop(AF_INET, &src, s, sizeof(s)), keys[i].sport);
+		sprintf(remote, ep_fmt, inet_ntop(AF_INET, &dst, d, sizeof(d)), ntohs(keys[i].dport));
+		printf("%-20s <-> %-20s %10llu\n", local, remote, counts[i]);
+	}
+}
+
+static void print_count_ipv6(int map_fd)
+{
+	static struct ipv6_flow_key keys[MAX_ENTRIES];
+	__u32 value_size = sizeof(__u64);
+	__u32 key_size = sizeof(keys[0]);
+	static struct ipv6_flow_key zero;
+	static __u64 counts[MAX_ENTRIES];
+	char s[INET6_ADDRSTRLEN];
+	char d[INET6_ADDRSTRLEN];
+	__u32 i, n = MAX_ENTRIES;
+	struct in6_addr src;
+	struct in6_addr dst;
+	char *ep_fmt = "[%s]#%d";
+	char remote[INET6_ADDRSTRLEN + 8];
+	char local[INET6_ADDRSTRLEN + 8];
+
+	if (dump_hash(map_fd, keys, key_size, counts, value_size, &n, &zero)) {
+		warn("dump_hash: %s", strerror(errno));
+		return;
+	}
+
+	for (i = 0; i < n; i++) {
+		memcpy(src.s6_addr, keys[i].saddr, sizeof(src.s6_addr));
+		memcpy(dst.s6_addr, keys[i].daddr, sizeof(src.s6_addr));
+		sprintf(local, ep_fmt, inet_ntop(AF_INET6, &src, s, sizeof(s)), keys[i].sport);
+		sprintf(remote, ep_fmt, inet_ntop(AF_INET6, &dst, d, sizeof(d)), ntohs(keys[i].dport));
+		printf("%-20s <-> %-20s %10llu\n", local, remote, counts[i]);
+	}
+}
+
+static void print_count(int map_fd_ipv4, int map_fd_ipv6)
+{
+	while (hang_on)
+		pause();
+
+	printf("\n%-25s %-25s %-10s\n", "LADDR:LPORT", "RADDR:RPORT", "RETRANSMITS");
+	print_count_ipv4(map_fd_ipv4);
+	print_count_ipv6(map_fd_ipv6);
+}
+
+static void print_events_header()
+{
+	printf("%-8s %-6s %-2s %-20s %1s> %-20s %-4s\n", "TIME", "PID", "IP",
```

Comment:
The output is slightly misaligned on my machine. Original tcpretrans.py has the same problem, so it might be my local values. But feels like it should be improved a bit, WDYT?

```
$ sudo ./tcpretrans
Tracing retransmits ... Hit Ctrl-C to end
TIME     PID    IP LADDR:LPORT          T> RADDR:RPORT          STATE
17:06:22 2142809 6  ::1:64056            R> ::1:8297             ESTABLISHED
17:06:22 2142809 6  ::1:64058            R> ::1:8297             ESTABLISHED
17:06:22 2142809 6  ::1:64060            R> ::1:8297             ESTABLISHED
17:06:22 2894834 6  ::1:37958            R> ::1:4331             ESTABLISHED
17:06:23 2142809 6  ::1:38622            R> ::1:16262            ESTABLISHED
17:06:23 2142809 6  ::1:38624            R> ::1:16262            ESTABLISHED
```

```
$ sudo ./tcpretrans -c
Tracing retransmits ... Hit Ctrl-C to end
^C
LADDR:LPORT               RADDR:RPORT               RETRANSMITS
[::1]#36560          <-> [::1]#9176                    1
[::1]#41082          <-> [::1]#9099                    1
[::1]#50392          <-> [::1]#3201                    1
[::1]#36744          <-> [::1]#9176                    1
```